interlocking/interlockingcontroller/trackcontroller.py

```diff
@@ -13,107 +15,111 @@ def __init__(self, interlocking, point_controller, signal_controller):
     def reset(self):
         for base_track_id in self.tracks:
             track = self.tracks[base_track_id]
-            for segment_id in track.state:
-                track.state[segment_id] = "free"
-
-    def set_route(self, route):
-        self.reserve_route(route)
-
-    def can_route_be_set(self, route):
-        segments = route.get_segments_of_route()
-        for track_base_id in segments:
-            track = self.tracks[track_base_id]
-            for segment_id in segments[track_base_id]:
-                if track.state[segment_id] != "free":
-                    return False
-        return self.overlap_controller.can_any_overlap_be_reserved(route)
+            for segment in track.segments:
+                segment.state = OccupancyState.FREE
+                segment.used_by = set()
+
+    def set_route(self, route, train_id: str):
+        self.reserve_route(route, train_id)
+
+    def can_route_be_set(self, route, train_id: str):
+        for segment in route.get_segments_of_route():
+            if segment.state == OccupancyState.FREE:
+                continue
+            if segment.is_only_used_by_train(train_id):
+                continue
+            print(f"{segment.segment_id} is used by other train")
+            return False
+        return self.overlap_controller.can_any_overlap_be_reserved(route, train_id)
 
     def do_two_routes_collide(self, route_1, route_2):
-        segments_of_route_1 = {x for v in route_1.get_segments_of_route().values() for x in v}
-        segments_of_route_2 = {x for v in route_2.get_segments_of_route().values() for x in v}
+        segments_of_route_1 = set(map(lambda seg: seg.segment_id, route_1.get_segments_of_route))
+        segments_of_route_2 = set(map(lambda seg: seg.segment_id, route_2.get_segments_of_route))
         if len(segments_of_route_1.intersection(segments_of_route_2)) > 0:
             return True
 
         overlaps_of_route_1 = route_1.get_overlaps_of_route()
         found_no_free_overlap_route_1 = True
         for overlap in overlaps_of_route_1:
-            overlap_segments_of_route_1 = {x for v in overlap.segments.values() for x in v}
-            if len(overlap_segments_of_route_1.intersection(segments_of_route_2)) == 0:
+            overlap_segments = set(map(lambda seg: seg.segment_id, overlap.segments))
+            if len(overlap_segments.intersection(segments_of_route_2)) == 0:
                 found_no_free_overlap_route_1 = False
                 break
```

Comment:
Doesn't the intersection of those two things need to be 0 for all overlaps in overlap_of_route_1? I think this should be the other way around. It should start as False and as soon as the intersection is > 0 once, it is set to True, because a collision has been found.
If so, it is the same below.

Reply:
No, there is only one overlap with an intersection of 0 necessary, to have non-colliding routes (for both routes). So if route 1 can set any overlap and route 2 can set any overlap, then the routes are non colliding. But I agree that this code is hard to understand and it fails when a route does not need any overlap, so I created a hopefully better understandable version.